hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowMetadataTableHistoryProcedure.scala

```diff
@@ -0,0 +1,293 @@
+                                   dataActiveTimeline: HoodieTimeline,
+                                   dataArchivedTimeline: Option[HoodieTimeline],
+                                   metadataActiveTimeline: HoodieTimeline,
+                                   metadataArchivedTimeline: Option[HoodieTimeline],
+                                   dataInstantInfoMap: Map[String, Map[HoodieInstant.State, HoodieInstantWithModTime]],
+                                   metadataInstantInfoMap: Map[String, Map[HoodieInstant.State, HoodieInstantWithModTime]]): Row = {
+
+    val dataInstant = findInstant(dataActiveTimeline, dataArchivedTimeline, instantTime)
+    val metadataInstant = findInstant(metadataActiveTimeline, metadataArchivedTimeline, instantTime)
+
+    val action = dataInstant.map(_.getAction).orNull
+    val state = dataInstant.map(_.getState.toString).orNull
+    val requestedTime = getFormattedDateForState(instantTime, HoodieInstant.State.REQUESTED, dataInstantInfoMap)
+    val inflightTime = getFormattedDateForState(instantTime, HoodieInstant.State.INFLIGHT, dataInstantInfoMap)
+    val completedTime = getFormattedDateForState(instantTime, HoodieInstant.State.COMPLETED, dataInstantInfoMap)
+
+    val mtAction = metadataInstant.map(_.getAction).orNull
+    val mtState = metadataInstant.map(_.getState.toString).orNull
+    val mtRequestedTime = getFormattedDateForState(instantTime, HoodieInstant.State.REQUESTED, metadataInstantInfoMap)
+    val mtInflightTime = getFormattedDateForState(instantTime, HoodieInstant.State.INFLIGHT, metadataInstantInfoMap)
+    val mtCompletedTime = getFormattedDateForState(instantTime, HoodieInstant.State.COMPLETED, metadataInstantInfoMap)
+
+    Row(
+      instantTime,
+      mtAction,
+      mtState,
+      mtRequestedTime,
+      mtInflightTime,
+      mtCompletedTime,
+      action,
+      state,
+      requestedTime,
+      inflightTime,
+      completedTime
+    )
+  }
+
+  private def findInstant(
+                           activeTimeline: HoodieTimeline,
+                           archivedTimeline: Option[HoodieTimeline],
+                           instantTime: String): Option[HoodieInstant] = {
+
+    val activeInstant = activeTimeline.getInstants.iterator().asScala
+      .find(_.requestedTime() == instantTime)
+
+    activeInstant.orElse {
+      archivedTimeline.flatMap { archived =>
+        archived.getInstants.iterator().asScala
+          .find(_.requestedTime() == instantTime)
+      }
+    }
+  }
+
+  private case class HoodieInstantWithModTime(state: HoodieInstant.State,
+                                              action: String,
+                                              requestedTime: String,
+                                              completionTime: String,
+                                              modificationTimeMs: Long
+                                             ) {
+    def getModificationTime: Long = modificationTimeMs
```

Comment:
iirc, scala case class variables can be directly accessed like `instantWithModTime.modificationTime`, and we don't need an extra getter